src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/ExtractPrimerTask.cpp

```diff
@@ -0,0 +1,187 @@
+
+    DocumentFormat *format = AppContext::getDocumentFormatRegistry()->getFormatById(BaseDocumentFormats::PLAIN_GENBANK);
+    SAFE_POINT_EXT(nullptr != format, setError(L10N::nullPointerError("Genbank Format")), );
+    QVariantMap hints;
+    QScopedPointer<Document> doc(format->createNewLoadedDocument(iof, settings.outputFileUrl, stateInfo, hints));
+    CHECK_OP(stateInfo, );
+
+    U2DbiRef dbiRef = AppContext::getDbiRegistry()->getSessionTmpDbiRef(stateInfo);
+    CHECK_OP(stateInfo, );
+
+    DNASequence productSequence = getProductSequence();
+    CHECK_OP(stateInfo, );
+    U2EntityRef productRef = U2SequenceUtils::import(stateInfo, dbiRef, productSequence);
+    CHECK_OP(stateInfo, );
+
+    QList<SharedAnnotationData> annotations;
+    int resultIndex = PCRPrimerDesignForDNAAssemblyTask::FRAGMENT_INDEX_TO_NAME.indexOf(settings.fragmentName);
+    
+    U2Region fragmentRegion = U2Region(0, productSequence.length());
+    if (!settings.backboneSequence.isEmpty() && settings.backboneLength != 0) {
+        SharedAnnotationData backboneAnnotationData(new AnnotationData());
+        QByteArray partToAdd = settings.backboneSequence.left(settings.backboneLength);
+        backboneAnnotationData->setStrand(U2Strand(U2Strand::Direct));
+        backboneAnnotationData->name = BACKBONE_ANNOTATION_NAME;
+        if (resultIndex % 2 == 0) {
+            productSequence.seq.prepend(partToAdd);
+            fragmentRegion.startPos =+ settings.backboneLength;
+            backboneAnnotationData->location->regions.append(U2Region(0, settings.backboneLength));
+        } else {
+            backboneAnnotationData->location->regions.append(U2Region(productSequence.length(), settings.backboneLength));
+            productSequence.seq.append(partToAdd);
+        }
+        annotations.append(backboneAnnotationData);
+    }
+    SharedAnnotationData fragmentAnnotationData(new AnnotationData());
+    fragmentAnnotationData->setStrand(resultIndex % 2 == 0 ? U2Strand(U2Strand::Direct) : U2Strand(U2Strand::Complementary));
+    fragmentAnnotationData->name = settings.fragmentName;
+    fragmentAnnotationData->location->regions.append(fragmentRegion);
+    annotations.append(fragmentAnnotationData);
+    U2SequenceObject *sequenceObject = new U2SequenceObject(productSequence.getName(), productRef);
+    sequenceObject->setWholeSequence(productSequence);
+    doc->addObject(sequenceObject);
+    AnnotationTableObject *annotationsTableObject = new AnnotationTableObject(productSequence.getName() + " annotations", dbiRef);
+    doc->addObject(annotationsTableObject);
+    annotationsTableObject->addObjectRelation(sequenceObject, ObjectRole_Sequence);
+    annotationsTableObject->addAnnotations(annotations, RESULT_ANNOTATION_GROUP_NAME);
+    result = doc.take();
+}
+
+Document *ExtractPrimerTask::takeResult() {
+    CHECK(nullptr != result, nullptr);
+    if (result->thread() != QCoreApplication::instance()->thread()) {
```

Comment:
The if` statement? You are not sure either the result in the main thread or not?

Reply:
This is for WD worker.